File: api/examples/result-modes/full-mode.ts
/**
 * Full Mode Example - TypeScript/JavaScript
 *
 * Use case: Deep research, CLI tools, comprehensive code review
 * Performance: <3000ms p95, ~50KB per result
 * Returns: All snippet fields + complete gitingest summaries + repository stats + dependencies
 */

const API_URL = process.env.API_URL || "https://govreposcrape-api-xxxxx-uc.a.run.app";

interface Dependency {
	name: string;
	version: string;
	type: "runtime" | "dev";
}

interface RepositoryStats {
	contributors: number;
	commits_last_month: number;
	open_issues: number;
	last_commit: string;
}

interface FullResult {
	repo_url: string;
	repo_org: string;
	repo_name: string;
	language: string;
	last_updated: string;
	similarity_score: number;
	github_link: string;
	metadata: {
		stars?: number;
		license?: string;
	};
	// Snippet fields
	snippet: string;
	snippet_file_path: string;
	snippet_line_range: string;
	context_lines_before: number;
	context_lines_after: number;
	codespaces_link: string;
	// Full mode fields
	gitingest_summary: string;
	full_file_context?: string;
	readme_excerpt?: string;
	repository_stats?: RepositoryStats;
	dependencies?: Dependency[];
}

interface SearchResponse {
	results: FullResult[];
	metadata: {
		query: string;
		limit: number;
		resultCount: number;
		duration: number;
	};
	mode: "minimal" | "snippets" | "full";
}

async function searchFull(query: string, limit: number = 2): Promise<SearchResponse> {
	try {
		const response = await fetch(`${API_URL}/mcp/search`, {
			method: "POST",
			headers: {
				"Content-Type": "application/json",
			},
			body: JSON.stringify({
				query,
				limit,
				resultMode: "full",
			}),
		});

		if (!response.ok) {
			const error = await response.json();
			throw new Error(`API Error: ${error.error.message}`);
		}

		return await response.json();
	} catch (error) {
		console.error("Search failed:", error);
		throw error;
	}
}

// Example usage
async function main() {
	try {
		const results = await searchFull("microservices architecture patterns", 2);

		console.log(
			`Found ${results.metadata.resultCount} repositories (took ${results.metadata.duration}ms)`,
		);
		console.log(`Mode: ${results.mode}\n`);

		results.results.forEach((repo, index) => {
			console.log(`${"=".repeat(80)}`);
			console.log(`${index + 1}. ${repo.repo_org}/${repo.repo_name} (${repo.language})`);
			console.log(`${"=".repeat(80)}`);
			console.log(`Score: ${repo.similarity_score.toFixed(2)}`);
			console.log(`Last Updated: ${repo.last_updated}`);
			console.log(`License: ${repo.metadata.license || "Unknown"}`);
			console.log(`Stars: ${repo.metadata.stars || "N/A"}`);

			if (repo.repository_stats) {
				console.log(`\nRepository Stats:`);
				console.log(`  Contributors: ${repo.repository_stats.contributors}`);
				console.log(`  Recent Commits: ${repo.repository_stats.commits_last_month}`);
				console.log(`  Open Issues: ${repo.repository_stats.open_issues}`);
				console.log(`  Last Commit: ${repo.repository_stats.last_commit}`);
			}

			console.log(`\nCode Snippet (${repo.snippet_file_path}, lines ${repo.snippet_line_range}):`);
			console.log(
				repo.snippet
					.split("\n")
					.map((line) => `  ${line}`)
					.join("\n"),
			);

			if (repo.readme_excerpt) {
				console.log(`\nREADME Excerpt:`);
				console.log(repo.readme_excerpt.substring(0, 200) + "...");
			}

			if (repo.dependencies && repo.dependencies.length > 0) {
				console.log(`\nKey Dependencies (${repo.dependencies.length} total):`);
				repo.dependencies.slice(0, 5).forEach((dep) => {
					console.log(`  - ${dep.name} ${dep.version} (${dep.type})`);
				});
				if (repo.dependencies.length > 5) {
					console.log(`  ... and ${repo.dependencies.length - 5} more`);
				}
			}

			console.log(`\nGitingest Summary (${repo.gitingest_summary.length} chars):`);
			console.log(repo.gitingest_summary.substring(0, 300) + "...\n");

			console.log(`Links:`);
			console.log(`  GitHub: ${repo.github_link}`);
			console.log(`  Codespaces: ${repo.codespaces_link}`);
			console.log("");
		});
	} catch (error) {
		console.error("Error:", error);
		process.exit(1);
	}
}

// Run if executed directly
if (require.main === module) {
	main();
}

export { searchFull, FullResult, SearchResponse, Dependency, RepositoryStats };

File: api/examples/result-modes/full_mode.py
#!/usr/bin/env python3
"""
Full Mode Example - Python

Use case: Deep research, CLI tools, comprehensive code review
Performance: <3000ms p95, ~50KB per result
Returns: All snippet fields + complete gitingest summaries + repository stats + dependencies
"""

import os
import requests
from typing import List, Dict, Optional, TypedDict


class Dependency(TypedDict):
    """Dependency entry"""
    name: str
    version: str
    type: str  # 'runtime' or 'dev'


class RepositoryStats(TypedDict):
    """Repository statistics"""
    contributors: int
    commits_last_month: int
    open_issues: int
    last_commit: str


class FullResult(TypedDict):
    """Full result format (comprehensive mode)"""
    repo_url: str
    repo_org: str
    repo_name: str
    language: str
    last_updated: str
    similarity_score: float
    github_link: str
    metadata: Dict[str, any]
    # Snippet fields
    snippet: str
    snippet_file_path: str
    snippet_line_range: str
    context_lines_before: int
    context_lines_after: int
    codespaces_link: str
    # Full mode fields
    gitingest_summary: str
    full_file_context: Optional[str]
    readme_excerpt: Optional[str]
    repository_stats: Optional[RepositoryStats]
    dependencies: Optional[List[Dependency]]


class SearchMetadata(TypedDict):
    """Search metadata"""
    query: str
    limit: int
    resultCount: int
    duration: int


class SearchResponse(TypedDict):
    """Complete search response"""
    results: List[FullResult]
    metadata: SearchMetadata
    mode: str


def search_full(query: str, limit: int = 2, api_url: Optional[str] = None) -> SearchResponse:
    """
    Search repositories using full mode.

    Args:
        query: Search query string
        limit: Maximum number of results (default: 2, full mode is slower)
        api_url: API endpoint URL (default: from API_URL env var)

    Returns:
        SearchResponse with full results including gitingest summaries

    Raises:
        requests.HTTPError: If API request fails
    """
    if api_url is None:
        api_url = os.environ.get('API_URL', 'https://govreposcrape-api-xxxxx-uc.a.run.app')

    response = requests.post(
        f"{api_url}/mcp/search",
        json={
            "query": query,
            "limit": limit,
            "resultMode": "full"
        },
        headers={"Content-Type": "application/json"},
        timeout=10  # Full mode may take longer
    )

    response.raise_for_status()
    return response.json()


def main():
    """Example usage"""
    try:
        results = search_full("microservices architecture patterns", limit=2)

        print(f"Found {results['metadata']['resultCount']} repositories "
              f"(took {results['metadata']['duration']}ms)")
        print(f"Mode: {results['mode']}\n")

        for i, repo in enumerate(results['results'], 1):
            print("=" * 80)
            print(f"{i}. {repo['repo_org']}/{repo['repo_name']} ({repo['language']})")
            print("=" * 80)
            print(f"Score: {repo['similarity_score']:.2f}")
            print(f"Last Updated: {repo['last_updated']}")
            print(f"License: {repo['metadata'].get('license', 'Unknown')}")
            print(f"Stars: {repo['metadata'].get('stars', 'N/A')}")

            if repo.get('repository_stats'):
                stats = repo['repository_stats']
                print(f"\nRepository Stats:")
                print(f"  Contributors: {stats['contributors']}")
                print(f"  Recent Commits: {stats['commits_last_month']}")
                print(f"  Open Issues: {stats['open_issues']}")
                print(f"  Last Commit: {stats['last_commit']}")

            print(f"\nCode Snippet ({repo['snippet_file_path']}, "
                  f"lines {repo['snippet_line_range']}):")
            for line in repo['snippet'].split('\n'):
                print(f"  {line}")

            if repo.get('readme_excerpt'):
                print(f"\nREADME Excerpt:")
                print(repo['readme_excerpt'][:200] + "...")

            if repo.get('dependencies'):
                deps = repo['dependencies']
                print(f"\nKey Dependencies ({len(deps)} total):")
                for dep in deps[:5]:
                    print(f"  - {dep['name']} {dep['version']} ({dep['type']})")
                if len(deps) > 5:
                    print(f"  ... and {len(deps) - 5} more")

            summary = repo['gitingest_summary']
            print(f"\nGitingest Summary ({len(summary)} chars):")
            print(summary[:300] + "...\n")

            print(f"Links:")
            print(f"  GitHub: {repo['github_link']}")
            print(f"  Codespaces: {repo['codespaces_link']}")
            print()

    except requests.HTTPError as e:
        print(f"API Error: {e}")
        if e.response is not None:
            error_data = e.response.json()
            print(f"Error code: {error_data['error']['code']}")
            print(f"Message: {error_data['error']['message']}")
        exit(1)
    except Exception as e:
        print(f"Error: {e}")
        exit(1)


if __name__ == "__main__":
    main()
